### PR TITLE
feat(server): soft-close thinking termination (qwen35 + gemma4)

### PR DESCRIPTION
**Pull request by easel**
## Summary

Adds soft-close thinking termination to qwen35 and gemma4 backends: a logit-ratio peek (`prob[</think>] / prob[chosen]`) lets the AR loop voluntarily inject the close sequence before the hard budget cap fires. Extends `BudgetHook` with split probe-vs-inject token ids, `soft_close_min_ratio`, and `soft_close_min_tokens` floor, and updates the thinking-budget spec to document the new `soft` taxonomy value and the operator/per-request dial. Backend-only; the matching CLI flags and request-envelope plumbing ship in #341.

## Files

- `server/src/common/model_backend.h` — extended `BudgetHook` struct (probe ids, min_ratio, min_tokens, helpers, `soft_close::` namespace)
- `server/src/qwen35/qwen35_backend.{cpp,h}` — soft-close peek wired into the AR loop
- `server/src/gemma4/gemma4_backend.cpp` — gemma4 port of the same pattern
- `server/test/test_server_unit.cpp` — soft-close unit tests (+533 lines)
- `docs/specs/thinking-budget.md` — `soft` taxonomy value, `soft_close_min_ratio` dial, per-request override semantics

## Dependencies

None. Soft-close is independent of the pflash/c2-gate stack — prior references in `qwen35_backend.cpp` to `c2_gate.h`, `req.use_transitive`, and `req.fa_window_override` have been dropped (those land via PR #274's pflash compression path, not this branch).

## Part of the PR #285 split

One of a series of tightly-scoped PRs splitting [Luce-Org/lucebox-hub#285](https://github.com/Luce-Org/lucebox-hub/pull/285) into reviewable chunks.

## Test plan

- [ ] Confirm `server/` builds standalone on this branch
- [ ] Run `test_server_unit` and confirm the new soft-close cases pass
- [ ] Stack with #341 to exercise end-to-end with `--think-soft-close-min-ratio` and a per-request override; verify the `stop_reason` taxonomy emits `soft` when the threshold clears before the hard cap
- [ ] Confirm legacy two-value (`natural` / `hard`) behavior is unchanged when `soft_close_min_ratio = 0.0` (default)
- [ ] Confirm `soft_close_min_tokens` floor prevents premature firing on short thinking blocks

## Risk

Low — soft-close is off by default (`min_ratio = 0.0`); existing hard-close and natural-close paths are unchanged when the dial is not engaged.

### CODE REVIEW

File: server/src/qwen35/qwen35_backend.cpp [hunk 1]
```diff
@@ -956,6 +947,79 @@ bool Qwen35Backend::do_ar_decode(int committed, int n_gen,
...
+            "inject[0]=%d (inject seq len %zu)\n",
+            committed_now, n_gen, remaining,
+            budget_hook.soft_close_min_ratio,
+            probe0, logits_row[probe0], logits_row[tok],
+            logits_row[probe0] - logits_row[tok],
+            std::log(budget_hook.soft_close_min_ratio),
+            tok, inject0, budget_hook.close_token_ids.size());
+        tok = inject0;
+        budget_close_started = true;
+        close_inject_pos = 1;
```
[the remark this answers is not in the document]
Comment (easel) [reply server/src/qwen35/qwen35_backend.cpp]: Fixed in 53ede18b. Set `close_inject_pos = 0` after soft fires (was 1). The continuation branch in `maybe_force_close` runs in the same AR step and reads `close_token_ids[close_inject_pos]`, so setting it to 1 caused the first close token to be silently overwritten with `close_token_ids[1]` on multi-token close sequences. With 0, the continuation no-op-rewrites tok=inject[0] and advances inject_pos to 1 for the next step. Applied symmetrically to gemma4 backend (same bug pattern).

File: server/src/gemma4/gemma4_backend.cpp [hunk 2]
```diff
@@ -262,6 +262,49 @@ bool Gemma4Backend::do_decode(int committed, int n_gen,
...
+    // than waiting for the hard think_max boundary. Mirrors qwen35's
+    // maybe_soft_close lambda — same BudgetHook fields, same soft_close::
+    // should_fire predicate from model_backend.h.
+    auto maybe_soft_close = [&](int32_t & tok,
+                                const float * logits_row,
+                                int committed_now) {
+        if (budget_close_started) return;
+        if (budget_hook.close_token_ids.empty()) return;
+        if (budget_hook.debug_thinking_logits) {
+            const int32_t close0 = budget_hook.close_token_ids.front();
```
[the remark this answers is not in the document]
Comment (easel) [reply server/src/gemma4/gemma4_backend.cpp]: Fixed in 53ede18b. Three issues addressed in gemma4 `maybe_soft_close`:

1. Now uses `budget_hook.soft_close_probe_token()` instead of `close_token_ids.front()` for the logit comparator — matches qwen35's probe-vs-inject split.
2. Added the `soft_close_min_tokens` floor gate before the `should_fire` check.
3. Also fixed the symmetric `close_inject_pos = 1` → `0` bug from comment #3358409731 (same root cause: gemma4's loop also calls `maybe_soft_close` then `maybe_force_close` in the same step).

Inject token is still `close_token_ids.front()` on fire, log line updated to report `probe0` and `inject[0]` separately.

File: server/src/qwen35/qwen35_backend.cpp [hunk 3]
```diff
@@ -1057,6 +1142,13 @@ bool Qwen35Backend::do_ar_decode(int committed, int n_gen,
             }
         }
 
+        // Soft check runs BEFORE hard-cap check. If soft fires, it sets
+        // budget_close_started=true so maybe_force_close's continuation
+        // branch handles steps 2..N of a multi-token close (and the
+        // remaining-check branch is skipped because the sequence is
+        // already started). If soft does not fire (disabled or threshold
+        // not met), maybe_force_close proceeds as today.
+        maybe_soft_close(next_tok, logits_buf.data(), committed);
```
[the remark this answers is not in the document]
Comment (easel) [reply server/src/qwen35/qwen35_backend.cpp:1649]: Fixed in 53ede18b. The GPU-argmax branch only reads 4 bytes (the argmax token id) and leaves `logits_buf` with stale data from the previous AR step, so calling `maybe_soft_close` with that buffer could misfire and emit garbage trajectory traces.

Added a guarded `ggml_backend_tensor_get` of the full logits row to CPU on the GPU-argmax path when either `soft_close_min_ratio > 0` or `debug_thinking_logits` is set. Zero-cost when soft-close is disabled — the D2H copy is skipped entirely (preserves the GPU-argmax fast path).
